app/ante.go

```diff
@@ -0,0 +1,234 @@
+	Cdc           codec.BinaryCodec
+	StakingKeeper stakingkeeper.Keeper
+	BankKeeper    bankkeeper.Keeper
+	IBCKeeper     *ibckeeper.Keeper
+}
+
+type MinCommissionDecorator struct {
+	sk         stakingkeeper.Keeper
+	bankkeeper bankkeeper.Keeper
+	cdc        codec.BinaryCodec
+}
+
+func NewMinCommissionDecorator(cdc codec.BinaryCodec, sk stakingkeeper.Keeper, bk bankkeeper.Keeper) MinCommissionDecorator {
+	return MinCommissionDecorator{cdc: cdc, sk: sk, bankkeeper: bk}
+}
+
+// getValidator returns the validator belonging to a given bech32 validator address
+func (min MinCommissionDecorator) getValidator(ctx sdk.Context, bech32ValAddr string) (stakingtypes.Validator, error) {
+	valAddr, err := sdk.ValAddressFromBech32(bech32ValAddr)
+	if err != nil {
+		return stakingtypes.Validator{}, sdkerrors.Wrap(sdkerrors.ErrInvalidAddress, bech32ValAddr)
+	}
+
+	val, found := min.sk.GetValidator(ctx, valAddr)
+	if !found {
+		return stakingtypes.Validator{}, sdkerrors.Register("ante", 12, "validator does not exist")
+	}
+
+	return val, nil
+}
+
+func (min MinCommissionDecorator) getTotalDelegatedTokens(ctx sdk.Context) sdk.Int {
+	bondDenom := min.sk.BondDenom(ctx)
+	bondedPool := min.sk.GetBondedPool(ctx)
+	notBondedPool := min.sk.GetNotBondedPool(ctx)
+
+	notBondedAmount := min.bankkeeper.GetBalance(ctx, notBondedPool.GetAddress(), bondDenom).Amount
+	bondedAmount := min.bankkeeper.GetBalance(ctx, bondedPool.GetAddress(), bondDenom).Amount
+
+	return notBondedAmount.Add(bondedAmount)
+}
+
+// Returns the projected voting power as a percentage (not a fraction)
+func (min MinCommissionDecorator) CalculateDelegateProjectedVotingPower(ctx sdk.Context, validator stakingtypes.ValidatorI, delegateAmount sdk.Dec) sdk.Dec {
+	validatorTokens := sdk.NewDecFromInt(validator.GetTokens())
+	totalDelegatedTokens := sdk.NewDecFromInt(min.getTotalDelegatedTokens(ctx))
+
+	projectedTotalDelegatedTokens := totalDelegatedTokens.Add(delegateAmount)
+	projectedValidatorTokens := validatorTokens.Add(delegateAmount)
+
+	return projectedValidatorTokens.Quo(projectedTotalDelegatedTokens).Mul(sdk.NewDec(100))
+}
+
+// Returns the projected voting power as a percentage (not a fraction)
+func (min MinCommissionDecorator) CalculateRedelegateProjectedVotingPower(ctx sdk.Context, validator stakingtypes.ValidatorI, delegateAmount sdk.Dec) sdk.Dec {
+	validatorTokens := sdk.NewDecFromInt(validator.GetTokens())
+	projectedTotalDelegatedTokens := sdk.NewDecFromInt(min.getTotalDelegatedTokens(ctx)) // no additional delegated tokens
+
+	projectedValidatorTokens := validatorTokens.Add(delegateAmount)
+
+	return projectedValidatorTokens.Quo(projectedTotalDelegatedTokens).Mul(sdk.NewDec(100))
+}
+
+func (min MinCommissionDecorator) AnteHandle(
+	ctx sdk.Context, tx sdk.Tx,
+	simulate bool, next sdk.AnteHandler,
+) (newCtx sdk.Context, err error) {
+	msgs := tx.GetMsgs()
+	minCommissionRate := sdk.NewDecWithPrec(5, 2) // 5% as a fraction
+	maxVotingPower := sdk.NewDecWithPrec(66, 1)   // 6.6%
```

Comment:
@kenta-elys we need to replace those by module parameters that can be changed with gov proposal.

Reply:
Got it.